ui/dashboards/src/context/DashboardProvider/layout-editing.ts

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  createPanelRef,
+  getPanelKeyFromRef,
+  GridDefinition,
+  GridItemDefinition,
+  LayoutDefinition,
+} from '@perses-dev/core';
+import { StateCreator } from 'zustand';
+import { Middleware } from './common';
+
+export interface LayoutEditorSlice {
+  layouts: LayoutDefinition[];
+
+  /**
+   * Given a LayoutItem location, returns the panel's unique key at that location.
+   */
+  getPanelKey: (layoutItem: LayoutItem) => string;
+
+  /**
+   * Add a panel with the specified key to an existing group.
+   */
+  addPanelToGroup: (panelKey: string, groupIndex: number) => void;
+
+  /**
+   * Move an existing Panel to a new panel group.
+   */
+  movePanelToGroup: (layoutItem: LayoutItem, newGroupIndex: number) => void;
+
+  /**
+   * Updates an existing panel group to, for example, change its display properties.
+   */
+  updatePanelGroup: (layout: LayoutDefinition, index?: number) => void;
+}
+
+/**
+ * The location of an item (e.g. a Panel) in layouts.
+ */
+export interface LayoutItem {
+  groupIndex: number;
+  itemIndex: number;
```

Comment:
maybe rename this to `panelIndex` since we're already referring to item as panel in other layout APIs

Reply:
Yeah, maybe the name `LayoutItem` is the wrong name here as well. 🤔